quickwit-indexing/src/actors/indexing_server.rs

```diff
@@ -0,0 +1,250 @@
+        self,
+    ) -> (ActorExitStatus, <IndexingServer as Actor>::ObservableState) {
+        self.handle.join().await
+    }
+}
+
+pub struct IndexingServer {
+    data_dir_path: PathBuf,
+    split_store_max_num_bytes: usize,
+    split_store_max_num_splits: usize,
+    metastore: Arc<dyn Metastore>,
+    storage_uri_resolver: StorageUriResolver,
+    indexing_pipeline_handles: BTreeMap<(String, String), ActorHandle<IndexingPipeline>>,
+}
+
+impl IndexingServer {
+    pub fn spawn(
+        indexer_config: IndexerConfig,
+        metastore: Arc<dyn Metastore>,
+    ) -> IndexingServerClient {
+        let universe = Universe::new();
+        let indexing_server = Self {
+            data_dir_path: indexer_config.data_dir_path,
+            split_store_max_num_bytes: indexer_config.split_store_max_num_bytes.get_bytes()
+                as usize,
+            split_store_max_num_splits: indexer_config.split_store_max_num_splits,
+            metastore,
+            storage_uri_resolver: quickwit_storage_uri_resolver().clone(),
+            indexing_pipeline_handles: Default::default(),
+        };
+        let (mailbox, handle) = universe.spawn_actor(indexing_server).spawn_async();
+        IndexingServerClient {
+            universe,
+            mailbox,
+            handle,
+        }
+    }
+
+    async fn observe_indexing_pipeline(
+        &mut self,
+        _ctx: &ActorContext<IndexingServerMessage>,
+        index_id: String,
+        source_id_opt: Option<String>,
+    ) -> anyhow::Result<HashMap<String, Observation<IndexingStatistics>>> {
+        let mut observations = HashMap::new();
+
+        if let Some(source_id) = source_id_opt {
+            let indexing_pipeline_handle = self
+                .indexing_pipeline_handles
+                .get(&(index_id.clone(), source_id.clone()))
+                .with_context(|| {
+                    format!(
+                        "Indexing pipeline `{}` for source `{}` does not exist.",
+                        index_id, source_id
+                    )
+                })?;
+            let observation = indexing_pipeline_handle.observe().await;
+            observations.insert(source_id, observation);
+        } else {
+            for ((_, source_id), indexing_pipeline_handle) in self
+                .indexing_pipeline_handles
+                .range((index_id.clone(), "".to_string())..)
+                .take_while(|((inner_index_id, _), _)| *inner_index_id == index_id)
+            {
+                let observation = indexing_pipeline_handle.observe().await;
+                observations.insert(source_id.clone(), observation);
+            }
+        }
+        Ok(observations)
+    }
+
+    async fn spawn_indexing_pipeline(
+        &mut self,
+        ctx: &ActorContext<IndexingServerMessage>,
+        index_id: String,
+        source_opt: Option<SourceConfig>,
+    ) -> anyhow::Result<()> {
+        let index_metadata = self.metastore.index_metadata(&index_id).await?;
```

Comment:
If the metastore takes too long to respond, the actor will timeout. Shall add a record progress right before?

Reply:
record_progress never hurt, yes.

We might even consider a protect_guard here.